File: backend/app/Models/appointment.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;

class Appointment extends Model
{
    use HasFactory;

    // Define the table name explicitly (optional if it matches the model name in plural form)
    protected $table = 'appointments';

    // The primary key for the table
    protected $primaryKey = 'appointment_id';

    // Disable auto-incrementing for the primary key (if necessary)
    public $incrementing = true;

    // The attributes that are mass assignable
    protected $fillable = [
        'patient_id',
        'doctor_id',
        'appointment_date',
        'reason_for_visit',
        'status',
        'duration',
        'appointment_type',
        'notes',
    ];

    // Set the date format for the model
    const CREATED_AT = 'created_at';
    const UPDATED_AT = 'updated_at';

    // Define relationships
    public function patient()
    {
        return $this->belongsTo(Patient::class, 'patient_id');
    }

    public function doctor()
    {
        return $this->belongsTo(Doctor::class, 'doctor_id');
    }

    // Set the date format for the appointment date
    protected $casts = [
        'appointment_date' => 'datetime',
    ];
}

File: backend/app/models/appointment.py
# app/models/appointment.py

from datetime import datetime
from sqlalchemy import (
    Column,
    Integer,
    DateTime,
    ForeignKey,
    Text,
    Enum,
    Index,
)

from sqlalchemy.orm import relationship
from app.database import Base 
from app.core.costants import AppointmentStatus

class Appointment(Base):
    __tablename__ = "appointments"

#----------------------------------------------
# Primary key 
#----------------------------------------------
    id = Column(Integer, primary_key=True, index=True)

#----------------------------------------------
# Foreign keys and other fields
#----------------------------------------------
    patient_id = Column(
        Integer, 
        ForeignKey("patients.id", ondelete="CASCADE"), 
        nullable=False,
        index=True
    )
    doctor_id = Column(
        Integer,
        ForeignKey("doctors.id", ondelete="CASCADE"),
        nullable=False,
        index=True
    )

#----------------------------------------------
# Appointment details
#----------------------------------------------
    appointment_date = Column(
        DateTime(timezone=True),
        nullable=False,
        index=True,
    )
    status = Column(
        Enum(AppointmentStatus, name="appointment_status"),
        nullable=False,
        default=AppointmentStatus.SCHEDULED,
        index=True,
    )
    
    reason = Column(Text, nullable=True)
    notes = Column(Text, nullable=True)

#----------------------------------------------
# Audit fields
#----------------------------------------------
    created_at = Column(
        DateTime(timezone=True), 
        nullable=False,
        default=datetime.utcnow,
    )
    updated_at = Column(
        DateTime(timezone=True),
        default=datetime.utcnow,
        onupdate=datetime.utcnow,
        nullable=False,
    )

#----------------------------------------------
# Relationships
#----------------------------------------------
    patient = relationship(
        "Patient", 
        back_populates="appointments",
        lazy="joined",
    )

    doctor = relationship(
        "Doctor", 
        back_populates="appointments",
        lazy="joined",
    )

    prescriptions = relationship(
        "Prescription", 
        back_populates="appointment", 
        cascade="all, delete-orphan",
        lazy="selectin",
    )

    medical_records = relationship(
        "MedicalRecord",
        back_populates="appointment",
        cascade="all, delete-orphan",
        lazy="selectin",
    )
 
#----------------------------------------------
# Table Indexes (Performance )
#----------------------------------------------
    __table_args__ = (
        Index(
            'ix_appointment_patient_date', 
            'patient_id', 
            'appointment_date'
        ),
        Index(
            'ix_appointment_doctor_date', 
            'doctor_id', 
            'appointment_date'
        ),
        Index(
            'ix_appointment_status_date', 
            'status', 
            'appointment_date'
        ),
    )

#----------------------------------------------
# String Representation
#----------------------------------------------
    def __repr__(self):
        return (
            f"<Appointment("
            f"id={self.id}, "
            f"patient_id={self.patient_id}, "
            f"doctor_id={self.doctor_id}, "
            f"appointment_date={self.appointment_date}, "
            f"status={self.status})>"
            f")>"
        )
    